src/main/java/io/anserini/collection/CommonCrawlWARCCollection.java

```diff
@@ -0,0 +1,746 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author mhoy@cs.cmu.edu (Mark J. Hoy)
+ */
+
+package io.anserini.collection;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.DataOutput;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Locale;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.zip.GZIPInputStream;
+
+/**
+ * A collection of WARC files from CommonCrawl (https://commoncrawl.org/the-data/get-started/#WARC-Format).
+ * This can be used to read the CommonCrawl WARC files
+ */
+public class CommonCrawlWarcCollection extends DocumentCollection<CommonCrawlWarcCollection.Document> {
+
+  public CommonCrawlWarcCollection(Path path) {
+    this.path = path;
+    this.allowedFileSuffix = Set.of(".warc.gz");
+  }
+
+  @Override
+  public FileSegment<CommonCrawlWarcCollection.Document> createFileSegment(Path p) throws IOException {
+    return new Segment(p);
+  }
+
+  /**
+   * An individual WARC in CommonCrawl.
+   */
+  public static class Segment extends FileSegment<CommonCrawlWarcCollection.Document> {
+    private static final byte MASK_THREE_BYTE_CHAR = (byte) (0xE0);
+    private static final byte MASK_TWO_BYTE_CHAR = (byte) (0xC0);
+    private static final byte MASK_TOPMOST_BIT = (byte) (0x80);
+    private static final byte MASK_BOTTOM_SIX_BITS = (byte) (0x1F);
+    private static final byte MASK_BOTTOM_FIVE_BITS = (byte) (0x3F);
+    private static final byte MASK_BOTTOM_FOUR_BITS = (byte) (0x0F);
+
+    protected DataInputStream stream;
+
+    public Segment(Path path) throws IOException {
+      super(path);
+      this.stream = new DataInputStream(new GZIPInputStream(Files.newInputStream(path, StandardOpenOption.READ)));
+    }
+
+    @Override
+    public void readNext() throws IOException, NoSuchElementException {
+      bufferedRecord = readNextWarcRecord(stream, Document.WARC_VERSION);
+    }
+
+    @Override
+    public void close() {
+      try {
+        if (stream != null) {
+          stream.close();
+        }
+        super.close();
+      } catch (IOException e) {
+        // There's really nothing to be done, so just silently eat the exception.
+      }
+    }
+
+    /**
+     * Our read line implementation. We cannot allow buffering here (for gzip
+     * streams) so, we need to use DataInputStream. Also, we need to account
+     * for Java's UTF8 implementation.
+     *
+     * @param in the input data stream
+     * @return the read line (or null if eof)
+     * @throws IOException if error encountered reading from stream
+     */
+    private static String readLineFromInputStream(DataInputStream in) throws IOException {
```

Comment:
We have code duplication here... either gather into an util class or use inheritance?